### PR TITLE
fix: find the `pnpapi` the `issuer` belongs to

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

- In some cases `enhanced-resolve` is outside the control of the `pnpapi` and can't require it
- The `issuer` might not belong to the same `pnpapi` as `enhanced-resolve`
- The `issuer` might not belong to a `pnpapi` at all and should use the `node_modules` resolution instead

Fixes https://github.com/arcanis/pnp-webpack-plugin/issues/32 for Webpack 5

**How did you fix it?**

Use `module.findPnpApi` to locate the correct `pnpapi` for the `issuer`, if a `pnpapi` isn't found continue the normal resolution

### CODE REVIEW

File: lib/PnpPlugin.js
Comment (merceyz) [inline lib/PnpPlugin.js:63]: I'm not quite sure how to let it use alternatives here

**Comment by alexander-akait** [comment]
I think better to fix it on yarn side?

**Comment by jgornick** [comment]
> I think better to fix it on yarn side?

@alexander-akait Can you elaborate on this how it could be fixed on the yarn side? Thank you for your input!

**Comment by merceyz** [comment]
It's not really a bug on our side, `require('pnpapi')` is supposed to return the `pnpapi` that the `issuer` (`enhanced-resolve` in this case) belongs to so it's working as intended

**Comment by jgornick** [comment]
@sokra @alexander-akait  ^^^ Bump 😄

File: lib/PnpPlugin.js
Comment (jgornick) [reply lib/PnpPlugin.js:63]: > I'm not quite sure how to let it use alternatives here

@sokra or @alexander-akait, if there's any insight into how we might be able to address a fix for this, I'd be more than happy to help! Thank you!

**Comment by jgornick** [comment]
@alexander-akait @sokra Just wanted to check-in again if you could help provide any insight on how to approach a fix for this issue? Again, I'm more than willing to help fix it too.

I've tried digging into this to figure out another way to help fallback to alternatives, but with the plugin system, I was getting lost as to how the flow of resolution works and where one might be able to intersect and fallback. I've also reached out to @arcanis to help provide any insight, and I think we're all kind of stuck.

Anything would be greatly appreciated! Thank you!

**Comment by alexander-akait** [comment]
Let's wait @arcanis answer, I am not against these changes

**Comment by sokra** [comment]
How to reproduce a scenario where you have packages that are yarn managed and packages that are non-yarn managed?

**Comment by jgornick** [comment]
> How to reproduce a scenario where you have packages that are yarn managed and packages that are non-yarn managed?

@sokra I'm sorry for not including that. Here's a link to a repo to reproduce that's using this patch as the fix: https://github.com/jgornick/yarn-webpack-node-modules-to-pnp

**Comment by jgornick** [comment]
> > How to reproduce a scenario where you have packages that are yarn managed and packages that are non-yarn managed?
> 
> @sokra I'm sorry for not including that. Here's a link to a repo to reproduce that's using this patch as the fix: https://github.com/jgornick/yarn-webpack-node-modules-to-pnp

@sokra Please let me know if there's anything else I can provide or help with 😄

**Comment by jgornick** [comment]
@sokra @alexander-akait ^^^ Bump again 😄 

Specifically looking for any information with how to resolve https://github.com/webpack/enhanced-resolve/pull/301#discussion_r701406073. I have a reproducible repository setup at https://github.com/jgornick/yarn-webpack-node-modules-to-pnp

Thank you!

**Comment by arcanis** [comment]
> Let's wait @arcanis answer, I am not against these changes

Those changes seem reasonable to me (note that @merceyz is a core contributor on Yarn 🙂).

**Comment by alexander-akait** [comment]
@jgornick Can you fix lint?

**Comment by jgornick** [comment]
I'll try to look at adding tests sometime today. 

Is it fair to say that out of the `PnpPlugin`, if it can't resolve a module that it should indeed [call the callback without any arguments](https://github.com/webpack/enhanced-resolve/pull/301/files#diff-484df5a516d012258b8eb71e51625ef132b1c571ef3756b7c45433f4365e2e69R64)? Or is there a better way we can tell enhanced-resolve to try fallbacks?

cc @sokra @alexander-akait

File: lib/PnpPlugin.js
Comment (jgornick) [reply lib/PnpPlugin.js:63]: @alexander-akait Thank you for approving, but the changes here won't work because of this question.

File: lib/ResolverFactory.js
Comment (merceyz) [reply lib/ResolverFactory.js:119]: It's checking if the PnP runtime has been initialized - https://yarnpkg.com/features/pnp#initializing-pnp

**Comment by jgornick** [comment]
@sokra and @alexander-akait ... There should be a commit coming from @merceyz that should fix this issue so that if PnP is available, but still can't resolve a module, it'll fallback to trying node_modules again.

The only question I have is how can I satisfy the code coverage issues, especially in [ResolverFactory](https://github.com/webpack/enhanced-resolve/pull/301/files#diff-7038853204325667ba3c6bff3b947028d7b2bf7a137cfc8074b87ca033dc0a67R121)? In order to test this, we'd have to mock a module so that the `findPnpApi` method is available. However, I don't see any other examples of that in the codebase, which makes me think it's an anti-pattern too 😄 

One option I was thinking of is what if we made the `findPnpApi` an `option` coming into the resolver factory so that it could be provided too? Please let me know what you think so I can finish this PR up 😄 

Thank you!

- Joe

File: lib/ResolverFactory.js
Comment (jgornick) [inline lib/ResolverFactory.js:435]: This is the heart of the change I made that allows the `PnpPlugin` to call the `callback` without any arguments so that it can fallback to the node_modules resolution plugin (`ModulesInHierachicDirectoriesPlugin`).

**Comment by jgornick** [comment]
For what it's worth @sokra, we've been using @merceyz's fix since October and haven't run into any issues. Is there any chance this can make it in?

**Comment by jgornick** [comment]
Bump @sokra and @alexander-akait 

My understanding is that we're only waiting on approvals? Any help or guidance to get this moving forward and merged into a release would be greatly appreciated!

**Comment by jgornick** [comment]
Bump @sokra and @alexander-akait 

Any chance we can get guidance on next steps here?

**Comment by jgornick** [comment]
@sokra, @alexander-akait or @vankop Can you provide any sort of feedback for this PR? As mentioned above, we've been using @merceyz's branch build of enhanced-resolve for over a year and fixed the particular issue described in this PR.

Again, if there's anything I can do to help merge this in, I'm more than willing to help! 

Please let me know either way.

**Comment by jgornick** [comment]
Bump @sokra, @alexander-akait or @vankop

Just wanting to know next steps for us here.

**Review by alexander-akait — APPROVED** [review]

**Comment by jgornick** [comment]
Thank you @alexander-akait for resetting this PR. I'm taking a look at the failing lint/tests.

**Comment by alexander-akait** [comment]
Tests failed, can someone take a look and fix it?

**Comment by alexander-akait** [comment]
@merceyz @jgornick Friendly ping :smile:

**Comment by jgornick** [comment]
@alexander-akait Thank you! I’ll make some time this weekend.

**Review by alexander-akait — APPROVED** [review]

File: lib/PnpPlugin.js
Comment (alexander-akait) [inline lib/PnpPlugin.js:70]: Just note - may we add test case there?

**Comment by jgornick** [comment]
@merceyz You're too quick 😛 Thank you very much for getting this squared away!

**Comment by alexander-akait** [comment]
@TheLarkInn Ready to review

**Comment by alexander-akait** [comment]
@merceyz Sorry again and again, let's rebase, thank you very much

File: package.json
Comment (alexander-akait) [inline package.json:4]: Let's avoid it here

File: package.json
Comment (alexander-akait) [inline package.json:14]: Why we `module` here?

File: package.json
Comment (jgornick) [reply package.json:4]: @merceyz I'm available to help get this through PR through now...

Do we need to specify `packageManager` in the `package.json` for `enhanced-resolve`? Does that help with the resolving logic?

File: package.json
Comment (jgornick) [reply package.json:14]: @merceyz I'm assuming we've specified the `module` main field here for resolvers that can utilize ESM? Do we need that though for `enhanced-resolve` and more specifically for this feature to make it through?

File: package.json
Comment (merceyz) [reply package.json]: Don't strictly need to but since the version of Yarn used by this project isn't documented I guessed and documented it.

File: package.json
Comment (merceyz) [reply package.json:13]: For the same reason `process` is there, it's not supported in browsers.

**Comment by jgornick** [comment]
@TheLarkInn and @alexander-akait Looks like this should be good to go.

**Comment by jgornick** [comment]
> Just fix merge conflicts and we are good to go!

@TheLarkInn Just wanted to check-in to make sure you're good with the changes here?

**Comment by jgornick** [comment]
@TheLarkInn and @alexander-akait: Bump 😄

**Comment by jgornick** [comment]
Sorry for the weekly bump @alexander-akait and @TheLarkInn, but just trying to get this through as I think we're finally all ready to go 😄

**Comment by jgornick** [comment]
Weekly bump @TheLarkInn 😄 

If we can get you to remove the changes requested flag, then we can at least get the workflows to run.

**Comment by jgornick** [comment]
Just checking in again on this PR to see if we can get it merged?

cc @TheLarkInn @alexander-akait

**Comment by jgornick** [comment]
👋 Hey, it's me again 😄 

Would love to get this merged so that we can stop referencing a custom build of enhanced-resolve.

cc @TheLarkInn @alexander-akait

**Comment by arcanis** [comment]
@TheLarkInn Is there anything still blocking the merge?

File: lib/ResolverFactory.js
Comment (alexander-akait) [inline lib/ResolverFactory.js:8]: Let's remove it from the top of file and put it inside `processPnpApiOption`

**Review by alexander-akait — COMMENTED** [review]
Tested and added some feedback

**Comment by alexander-akait** [comment]
I'm a little concerned about performance, we need improve it

**Review by alexander-akait — APPROVED** [review]

**Comment by jgornick** [comment]
Thank you @alexander-akait!

**Pull request merged**
